### PR TITLE
Use the new header API

### LINKED ISSUE #856: addreplacerg overwrite_all doesn't overwrite the old RG tags
I wanted to use samtools(v1.8) addreplacerg to replace the RG tags in a cram file.
My old RG tag is 

    @RG     ID:NG01Ga       LB:NG01Ga       PL:ILLUMINA     SM:NG01Ga
    @RG     ID:NG01Gb       LB:NG01Gb       PL:ILLUMINA     SM:NG01Gb

So, I run this command in RedHat system.

    samtools addreplacerg -m overwrite_all -r '@RG\tID:NG01G\tLB:NG01G\tPL:ILLUMINA\tSM:NG01G' --reference ASM.fa -O CRAM -o NG01G_reheader.cram -@ 5 NG01G.cram

Then, in the new cram file, there're three RG tags, the new one is appended, not replace the old.

    @RG     ID:NG01Ga       LB:NG01Ga       PL:ILLUMINA     SM:NG01Ga
    @RG     ID:NG01Gb       LB:NG01Gb       PL:ILLUMINA     SM:NG01Gb
    @RG     ID:NG01G        LB:NG01G        PL:ILLUMINA     SM:NG01G

But, I have used the overwrite mode.
Is this normal?

### LINKED ISSUE #1079: make error
#### Is your feature request related to a problem? Please specify.
I got an error when compile.
OS : Mac Mojave 
./configure worked; however, when I tried to "make", I keep getting an error message as below.

Donghoons-MacBook-Pro-2015:samtools hoon$ make
gcc -g -O2 -I. -I/Library/SAMTools/htslib-1.9/ -I./lz4  -c -o bam_plcmd.o bam_plcmd.c
bam_plcmd.c:888:60: error: expected ')'
                                    fprintf(pileup_fp, "%" PRId64 "", tag_i);
                                                           ^
bam_plcmd.c:888:44: note: to match this '('
                                    fprintf(pileup_fp, "%" PRId64 "", tag_i);
                                           ^
1 error generated.
make: *** [bam_plcmd.o] Error 1

I installed HTSlib-1.9 under /Library/SAMTools/htslib-1.9 and I configured as above.
I would appreciate any help for this problem.
Thanks
Hoon

### PR DESCRIPTION
**Pull request by valeriuo**
Depends on htslib PR https://github.com/samtools/htslib/pull/812
Fixes #856
Fixes #1079

### CODE REVIEW

**Comment by daviesrob** [comment]
Updated to the header API as finally deployed.  Most of the original commits were just chasing changes in the API as it developed, hence squashed as not very useful.

**Comment by jmarshall** [comment]
Note for when merging or further rewriting these commits: the first one has inadvertently deleted _test/split/.gitignore_.

**Comment by valeriuo** [comment]
Thanks for spotting that. Fixed and rebased.

**Comment by valeriuo** [comment]
1161e5b027cd9aa356d23961f067922937253c5d fixes #856

File: bam_addrprg.c [hunk 1]
```diff
@@ -354,6 +354,11 @@ static bool init(const parsed_opts_t* opts, state_t** state_out) {
             fprintf(stderr, "[init] Error adding RG line with ID:%s to the output header.\n", opts->rg_id);
             return false;
         }
+        if (opts->mode == overwrite_all &&
+            -1 == sam_hdr_keep_line(retval->output_header, "RG", "ID", opts->rg_id)) {
```
Comment (jmarshall) [inline bam_addrprg.c]: Thanks for fixing this.

It's a bit surprising that a function called `sam_hdr_keep_line` has the effect of removing lines! Maybe this API function could be renamed to `sam_hdr_keep_only_line` or similar?

Comment (valeriuo) [reply bam_addrprg.c]: Renamed to `sam_hdr_remove except`.

**Comment by valeriuo** [comment]
Fixes #1079

**Comment by daviesrob** [comment]
Rebased so all commits build against current HTSlib.

Upgraded `bam_split` `count_RG()` function to use the new header API.

`bam_smpl_add()` and `get_rg_sample()` also do some header text parsing.  I'll probably them separately as they are a bit more complicated.

**Pull request merged**